Add a lightweight domain "gathering" system

konklone opened the pull request:
This adds a `./gather` CLI, similar to the `./scan` CLI, which focuses on gathering hostnames from different sources.

For example, one might run:

```bash
./gather censys --suffix=.gov --sort
```

And this would gather a page of `.gov` domain data from the Censys API, using the `gather(suffix, options)` method inside `gatherers/censys.py`.

A `--suffix` is currently required. Parent (second-level) domains are excluded by default, as are `www` subdomains of parent domains. Include them with `--include-parents`. You can also pass a CSV path or URL as `--parents`, which will cause the column of domains in that CSV to be treated as a whitelist of base domains, and any subdomains that don't match the whitelist will be excluded. This is all documented in the README.

This has some different properties than the scanning process:

* Each "gatherer" is expected to have one long monolithic process, rather than making a request per-domain, so there's no facility for parallelization. 
* It only runs one "gatherer" at a time, rather than allowing multiple gatherers to be done in sequence. Maybe this will change later, but gatherers are expected to require more sophisticated command line flags, and the prospect of clashing flags is much higher than with scanners.
* It currently requires a "suffix", such as `.gov`, that is expected to be applied by each gatherer in whatever way makes sense for it. For example, the `censys` gatherer passes the suffix to the Censys API in order to limit the pages of data it needs to walk through.

And some similarities:

* Uses `cache/` and `results/`, just like `./scan` does. The output from `./gather` will clobber the output from `./scan`, just like repeated scans will clobber each other.
* CLI options parsing, logging, and custom output directories (via `--output`) are all shared between both CLI tools. In-place CSV sorting code is also shared, now in `utils.py`.
* A `meta.json` file is created in `results/`, with the command and timestamps of the gathering process.

Right now, the only "gatherer" is `censys`. I'm planning to imminently add another one for data from the Digital Analytics Program, which may just end up being a `csv` gatherer that expects a URL to a remote CSV (rather than being something specific to DAP).

I've added documentation to the README about it. The `scripts/` folder remains, with a couple files in it, but it's sort of an unmaintained space.

Some TODOs that come to mind for later:

* Move `scanners/utils.py` to some place neutral, since it's now being used by both scanners and gatherers.
* Make some sort of effort to fix up the `censys` scanner so that you don't have to figure out the number of pages in advance. This is a [limitation of the `censys` Python library](https://github.com/Censys/censys-python/issues/10). Since that library is not stable or actively maintained, and it currently doesn't add a lot of value, we may want to roll our own lightweight one that does just enough to gather hostnames.
* Add some caching logic to `censys` so that it can cache pages of results (useful during dev/testing).

konklone commented:
I added a `url` gatherer that fetches a remote CSV URL and applies the `--parents` and `--suffix` filters to it appropriately, as well as excluding second-level domains by default. 

I'm using this personally to fetch [tracked .gov hostnames](https://analytics.usa.gov/data/live/sites.csv) from the [Digital Analytics Program](https://analytics.usa.gov).

konklone commented:
I vendor'd the [`censys-python`](https://github.com/Censys/censys-python) project so that no extraordinary installation/setup steps were required to use the `censys` gatherer. I've updated the LICENSE file to acknowledge the exception to the repo's license terms, and excluded `vendor/censys` from `flake8` checks in Travis.

alex commented:
Is there a reason not to just put a `git+https://...` in your
requirements.txt?

On Sat, Sep 3, 2016 at 10:51 PM, Eric Mill <notifications@github.com> wrote:

> I vendor'd the censys-python <https://github.com/Censys/censys-python>
> project so that no extraordinary installation/setup steps were required to
> use the censys gatherer. I've updated the LICENSE file to acknowledge the
> exception to the repo's license terms, and excluded vendor/censys from
> flake8 checks in Travis.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/18F/domain-scan/pull/85#issuecomment-244581269>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAADBOe8fGbHOZ1gHyaQiqRqCBp6R_WHks5qmjIpgaJpZM4J0cJD>
> .
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: D1B3 ADC0 E023 8CA6

konklone commented:
@alex That is a **way** better idea. I've used that before, but never actually set it up myself, so I didn't think of it.

I just added some commits undoing the vendor-ing, and added the git repo to `requirements.txt`

konklone commented:
I added a commit that uses a new feature in the censys-python library to fetch the number of pages to request from the Censys API per-query, which allows for much more natural behavior.

This will now Just Work and fetch all known `.gov` hostnames from the Censys API:

```bash
./gather censys --suffix=.gov
```

blkperl commented on requirements.txt:
```diff
@@ -6,3 +6,6 @@ strict-rfc3339
 python-dateutil
+
+# to support censys gatherer
+git+https://github.com/Censys/censys-python
```
looks like its on pypi now. Uploaded: 2016-09-04

https://pypi.python.org/pypi/censys/0.0.6

konklone replied:
Yes they did! Just updated the PR to use 0.0.6 instead of pulling from git.

blkperl commented on gatherers/censys.py:
```diff
@@ -0,0 +1,86 @@
+
+# Register a (free) Censys.io account to get a UID and API key.
+uid = os.environ.get("CENSYS_UID", None)
+api_key = os.environ.get("CENSYS_API_KEY", None)
```
You should file an issue against censys-python to do this automatically. Then leaving a comment linking to the issue.

konklone replied:
I don't desire it from the censys-python library, personally -- but if you do, you should definitely open an issue.

konklone commented:
I added a few commits to:

* De-dupe hostnames across all scanners, in the parent `gather` script. This means the `censys` gatherer is de-duping redundantly, but I didn't feel like de-duping de-duping.
* Allows the `url` gatherer to also take in a local path (if it doesn't start with `http:` or `https:`). I didn't want to name the gatherer `csv` or `file`, which are already core Python libs/names, I'm just leaving it as `url`.
* Updated `censys-python` to `0.0.6`, since the PyPi package was updated since my last comment.

With that, I'm merging, and beginning to integrate this into https://pulse.cio.gov.
